### PR TITLE
Improve error reporting using entry point attribute in a script

### LINKED ISSUE #1103: `Tracking` Error reporting improvenets
Related to #1102, this issue could act as an overall list for all the issues relating to error reporting from the compiler. I'll start the ball rolling by adding the ones mentioned in that issue plus a few more.

### Up for grabs!
- [ ] #12341
- [x] #1160
- [x] #2739
- [x] #2799
- [x] #2828
- [x] #2321
- [x] #3302
- [x] #3765
- [ ] #4233
- [ ] #4232
- [ ] #5476
- [ ] #5259
- [ ] #8127
- [ ] #10167
- [ ] #6873

### More discussion / on hold
- [ ] #13014
- [ ] #13013
- [ ] #13012
- [x] #1107
- [x] #1120
- [x] #1136
- [x] #1161
- [x] #1348
- [x] #1384
- [x] #1429
- [x] #2277
- [x] #1826
- [x] #1511
- [x] #1442
- [ ] #5236
- [ ] #5475
- [ ] #6687
- [ ] #6712
- [ ] #6717
- [x] #6872
- [ ] #11234
- [ ] #12586
- [ ] #10564
- [ ] #11445
- [x] #11468

### Implemented!
- [x] #1102
- [x] #1104
- [x] #1105
- [x] #1110
- [x] #1122
- [x] #1127
- [x] #1129
- [x] #1157
- [x] #1162
- [x] #1170
- [x] #1214
- [x] #1275
- [x] #1203
- [x] #1276
- [x] #1112
- [x] #1108
- [x] #1232
- [x] #1280
- [x] #1834
- [x] #1109
- [x] #2561
- [x] #1218
- [x] #326
- [x] #1905
- [x] #1933
- [x] #1430
- [x] #2268
- [x] #569
- [x] #4228
- [x] #5114
- [x] #4982
- [x] #8091
- [x] #1121
- [x] #2201 (more discussion?)
- [x] #2307 (should still be open for more "complete" fix)
- [x] #920
- [x] #1431

### Rejected
- [x] #1106
- [x] #1128

## Contributing

If you have ideas for other error messages, please adhere to the following standard for new issues: -
1. Title the issue "Improve Error Reporting: <issue>"
2. Each issue should have the following elements: -
   
    a. What - reproduction of the error through a code snippet and compiler error / warning.
    b. Why - why the current compiler error / warning is insufficient.
    c. How - how you would improve the current error / warning with an alternative example.

### LINKED ISSUE #1431: Improve error reporting: using EntryPoint attribute in an F# script
## What

Running the code:

``` fsharp
[<EntryPoint>]
let main argv =
    printfn "Hello, world!"   
    0
```

in FSI produces the output
`val main : argv:string [] -> int`
but doesn't actually invoke the function.
## Why

Some beginners may be unsure about how F# scripts and FSI work and may run some code from an application which has an `EntryPoint` into FSI. This happened recently to a former colleague.
## How

Using the `EntryPoint` attribute in FSI should produce a compiler warning indicating that it won't be invoked automatically.

### PR DESCRIPTION
**Pull request by edgarfgp**
Fixes #1431

I took the warning text based on the discussion on #1431  . But Feel free to suggest a new one if you think is better :)

### CODE REVIEW

File: src/Compiler/Checking/PostInferenceChecks.fs [hunk 1]
```diff
@@ -2101,6 +2101,7 @@ let CheckModuleBinding cenv env (TBind(v, e, _) as bind) =
     let isExplicitEntryPoint = HasFSharpAttribute g g.attrib_EntryPointAttribute v.Attribs
     if isExplicitEntryPoint then 
         cenv.entryPointGiven <- true
+        warning (Error(FSIstrings.SR.fsiEntryPointWontBeInvokedAutomatically(v.DisplayName, v.DisplayName, v.DisplayName), v.Range))
```
Comment (vzarytovskii) [inline src/Compiler/Checking/PostInferenceChecks.fs:2104]: This will affect non-scripts too, won't it?

**Review by vzarytovskii — COMMENTED** [review]
I also think that @0101 started working on it already.

File: src/Compiler/Checking/PostInferenceChecks.fs (hunk 1 above)
Comment (edgarfgp) [reply src/Compiler/Checking/PostInferenceChecks.fs:2104]: I could not run the test locally for some reason :( So I used the CI to see how this affects the rest of the test .

Comment (edgarfgp) [reply src/Compiler/Checking/PostInferenceChecks.fs:2104]: @vzarytovskii  I’m not sure where to raise the warning only for the scripts TBH . So if someone is already on it happy to close this PR

Comment (edgarfgp) [reply src/Compiler/Checking/PostInferenceChecks.fs:2104]: Sure. go for it :) . Will learn from your PR

**Pull request closed without merging**